Handle 'PermissionError' and Ensure valid filenames

Raqeeb27 opened the pull request:
### Overview

This pull request addresses potential `PermissionError` exceptions during the hiding and recovery processes when a file is in use by another process. It also ensures that only files with valid filenames are processed.

### Changes

* **`hiding.py`:**
    * Implemented checks to validate filenames (ensuring they have both a name and an extension).
    * Added error handling for `PermissionError` during the hiding process.
* **`recovery.py`:**
    * Implemented checks to validate filenames (ensuring they have both a name and an extension).
    * Added error handling for `PermissionError` during the recovery process.

### Purpose

`shutil.move` copies then deletes the source file and doesn't check if a file is in use before copying. If another process uses the source during hiding or recovery, it causes a `PermissionError` and the source is copied to destination but the source is not deleted, causing an `OSError`. To fix this `PermissionError` and support cross-drive operations, this pull request uses a combination of `os.rename` and `shutil.move`. `os.rename` is tried first and if it fails then file is in use or it's a cross-drive move. For cross-drive moves, `shutil.move` is used.

### Testing

* Confirmed that the application now only processes files with valid filenames (containing both a name and an extension).
* Verified that `PermissionError` exceptions are gracefully handled during both hiding and recovery processes when a file is in use.
* Tested and confirmed successful cross-drive hiding and recovery operations.

### Notes

Please review these changes and provide a feedback before merging.

Raqeeb27 committed:
fix: Handle 'PermissionError' and ensure valid file names

Raqeeb27 committed:
fix: Ensure shortcuts with valid filnames and handle 'PermissionError'

hanXen commented:
### Question Regarding Recovery Logic

Hi there,

I noticed this snippet in the recovery logic:

```python
while original_file.strip().endswith(" - Copy"):
    original_file = original_file.removesuffix(" - Copy")
```

Could you help me understand the reasoning behind it?

1. Is there a real-world scenario where `.lnk` (shortcut) files get copied?
2. If supporting copied shortcuts is indeed important, would it make sense to handle other locale-specific suffixes (e.g., `" - Kopie"`, `" - Copia"`, etc.) rather than hard-coding only `" - Copy"`?

In my current thinking, if `.lnk` files are indeed copied and this functionality is essential, the implementation should be revised to be more reliable. Instead of simply removing the `- Copy` suffix, it would be better to make the logic adaptable to different locales, ensuring it handles cases universally and robustly.

Clarifying these points will help us determine the best path forward. Thanks!

hanXen commented on hiding.py:
```diff
@@ -123,7 +132,14 @@ def make_shortcut(file_path: str, ext_icon_dict: dict[str, str],
 
     try:
-        shutil.move(file_path, hidden_file_path)
+        try:
```
The logic for moving files is repeated across both `hiding.py` and `recovery.py`. I suggest creating a `move_file()` function in `common_utils.py` to centralize and streamline this operation.

Raqeeb27 replied:
Yeah definitely. Thank you.

hanXen commented on hiding.py:
```diff
@@ -86,6 +86,15 @@ def make_shortcut(file_path: str, ext_icon_dict: dict[str, str],
+    find_ext = os.path.basename(file_path).rfind(".")
+    if find_ext == -1:
+        print(f"[-] Failed to hide {file_path} : Missing extension in filename.")
+        return None
```
Is the intention behind handling `find_ext == 0` to avoid supporting hidden files like `.gitignore`? If so, I think the following approach would be better.

Suggestion: Consider using `os.path.splitext()` instead of `rfind` for extracting the file extension. This approach directly separates the file name and extension, making the logic clearer and reducing the reliance on index checks like `0` or `-1`.

Example code:
```python
if not os.path.splitext(original_file)[1]:
    print(f"[-] Failed to hide {file_path} : Missing extension in filename.")
    return None
if original_file.startswith("."):
    print(f"[-] Failed to hide {file_path} : Files starting with a dot (.) are not supported for hiding.")
    return None
```

Raqeeb27 replied:
My focus was on the precise print statements regarding the error. I tried os.path.splitext() but it returns an empty extension when the filename is just an extension itself (like '.txt') or even without any dot ('txt'). That made me choose .rfind() for catching the error efficiently.

hanXen replied:
How about writing the validation logic like this?

``` python
if original_file.startswith("."):
    print(f"[-] Failed to hide {file_path}: Files starting with a dot (.) are not supported for hiding.")
    return None
if not os.path.splitext(original_file)[1]:
    print(f"[-] Failed to hide {file_path}: Missing extension in filename.")
    return None
```

With this order:
1. Any name beginning with a dot (e.g., .txt) is immediately rejected.
2. Then, filenames lacking a “.” extension (e.g., txt) trigger the “Missing extension” error.

This way, '.txt' gets caught by the first check, and 'txt' gets caught by the second.

This approach makes the code's intent much clearer and avoids ambiguity in how filenames are handled.

If this approach seems reasonable, please consider including it in the next PR version, or I can proceed with the refactoring.

hanXen commented on recovery.py:
```diff
@@ -36,18 +36,32 @@ def recovery(hidden_file: str, mapping_dict: dict[str, str],
 
+        find_ext = os.path.basename(original_file).rfind(".")
+        if find_ext in [0, -1]:
+            raise NameError(f"Invalid filename {original_file} : Please rename it to valid filename.")
```
Same as the comment on line 90 of `hiding.py`.
And, the NameError here seems misaligned; ValueError might be more suitable for filename validation.

Raqeeb27 replied:
Yes, got it! I'll change it to `ValueError`

hanXen reviewed (commented):
This is an implementation-focused review summary. It covers the following points:
- Filename Validation
- File Move Logic Duplication
- Error Handling in Recovery Logic

Raqeeb27 commented:
* Regarding the '- Copy' suffix, I noticed that copying a link file within the same directory appends this suffix repeatedly. During the recovery process, it fails to reconstruct the original filename and removing the suffix fixes it instantly.
* While this was an initial quick fix, a more robust solution is definitely needed to handle it efficiently.

I'll try to make the changes to the files.

**Thank you for the detailed review**

Raqeeb27 commented:
Hi there,

I've addressed all the issues discussed in this PR.

Regarding the next steps, would you prefer to merge this PR, to make further updates to it, or close this one and open a new PR?

Given below are the extensive changes across multiple files, I believe it might be easier to manage and review if I open separate PRs, one for each file. This would allow for more focused discussions and potentially faster integration.

* `security_utils.py`: Implemented minimum password requirements for verified passwords, including a maximum number of password retry attempts.
* `init_db.py`: Minor changes to ensure proper integration with the `security_utils.py` module.
* `install.ps1`: Enhanced echo statements for improved user experience and integration with `init_db.py`.
* `set-rightclick.bat` and `remove-rightclick.bat`: Added 2 right-click menu options to recover link files within the current folder or recursive search in subfolders as well.
* `common_utils.py`: Added a `move_file()` method.
* `hiding.py`: Addressed the previously discussed issues and improved argument validation.
* `recovery.py`: Significant updates and new features:
    * Added functions: `process_link_file()`, `search_lnk_files()`, `find_lnks_with_hash()`, `recover_valid_found_links()`.
    * Addressed the issues discussed in the previous PR within the `recovery()` method.
    * Enhanced the `main` logic to reduce redundancy.
    * Improved argument validation.
    * Separated the `--all` and `--testbed` functionalities.
    * Introduced a `--dir` argument to recover files within the current folder and --recursive argument to recover files in subfolders.

With the issues in this PR being addressed, merging it would allow for easy patching of these changes.

Thank you for your active participation in merging PRs and for providing such detailed reviews on the issues I encountered in this PR.

**I look forward to making further contributions to this project.**

hanXen commented:
> * Regarding the '- Copy' suffix, I noticed that copying a link file within the same directory appends this suffix repeatedly. During the recovery process, it fails to reconstruct the original filename and removing the suffix fixes it instantly.
> * While this was an initial quick fix, a more robust solution is definitely needed to handle it efficiently.
> 
> I'll try to make the changes to the files.
> 
> **Thank you for the detailed review**

Copying a `.lnk` file does not mean the original file itself is duplicated. Therefore, I think it’s unlikely for `.lnk` files to be copied within the scope of this project.

Considering `.lnk` file copies is a thoughtful approach, but the current implementation assumes only a specific locale, which makes it less refined.

It might be better to revisit this feature later with a more comprehensive solution that can handle broader cases. Given the pressing priorities, I suggest excluding this functionality for now and considering it in the future.

Thank you for considering these detailed aspects of the project.

hanXen commented:
It seems you've made significant updates based on this PR. I agree that merging this PR will provide convenience for future patches. Therefore, I will proceed with merging the modifications included in this PR.

However, I kindly ask you to consider the following points I recently commented on for the next PR:  
- Transitioning from `rfind()` to `os.path.splitext()` for improved clarity and robustness.  
- Excluding the consideration of copied `.lnk` files.  

While submitting PRs by file is a valid approach, structuring commits/PRs based on functional changes would be even more helpful for tracking modifications.  
For example, grouping the addition of `move_file()` in `common_utils.py` and its application in `hiding.py` and `recovery.py` into a single commit or creating file-based commits within one PR would improve traceability.  

Additionally, for refactoring elements such as adding `move_file()`, it would be helpful to use a commit convention like `refactor:` to explicitly indicate that the changes are part of a code refactor.  
For changes like excluding the logic for handling `- Copy`, creating a separate commit with a descriptive commit message would make it easier to track the modifications.
Clear and detailed commit messages make it easier to track and understand what changes have been made during the development process.

I hope you understand that my responses might be delayed due to my current workload.  

Thank you for your dedication to this project. I’m learning a lot from your contributions.

Raqeeb27 commented:
I understand that link files are not typically copied directly. However, given that our project hides files within system paths, a user wishing to copy their files would face the inconvenience of recovering each file, manually copying it, and then hiding them again. Therefore, I believe that considering the copying scenario falls within the scope of this project to enhance user experience.

I plan to introduce a feature where, when using the `--all` or `--dir` arguments, the hidden files will be copied to the link file paths that share the same hashes. This would enable users to recover the hidden file to all locations where the corresponding link files have been copied.

To address the limitations of the previous approach, I've updated the code by introducing the `validate_extension()` function in `common_utils.py`. This function now efficiently removes any suffixes in the filenames and ensures that the filenames are updated to remove any leading or trailing whitespace. This prevents issues where `os.path.splitext` might incorrectly interpret a filename with leading spaces before the extension (e.g., `' .txt'`) as having a valid extension.

```python
directory = os.path.dirname(file_path)
file_path = os.path.join(directory, os.path.basename(file_path).strip())
ext = os.path.splitext(file_path)[1].split(" ")[0]
if not ext:
    print(f"Error in filename {file_path}: Invalid or Missing file extension.")
    sys.exit(1)
else:
    return file_path[:file_path.rfind(ext) + len(ext)]
```

However, this snippet assumes that there exists a whitespace after the extension. Further updates are required to address the limitations of this approach.

**Thank you for merging this pull request.** I will indeed make clear commits for each file based on the changes made to it.
